query/executor.go

```diff
@@ -289,6 +292,98 @@ func (e *Executor) WithLogger(log *zap.Logger) {
 	e.TaskManager.Logger = e.Logger
 }
 
+func initQueryLogWriter(log *zap.Logger, e *Executor, path string) (*os.File, error) {
+	logFile, err := os.OpenFile(path, os.O_CREATE|os.O_RDWR|os.O_APPEND, 0644)
+	if err != nil {
+		e.Logger.Error("failed to open log file", zap.Error(err))
+		return nil, err
+	}
+
+	existingCore := log.Core()
+
+	encoderConfig := zap.NewProductionEncoderConfig()
+
+	fileCore := zapcore.NewCore(
+		zapcore.NewJSONEncoder(encoderConfig),
+		zapcore.Lock(logFile),
+		zapcore.InfoLevel,
+	)
+
+	newCore := zapcore.NewTee(existingCore, fileCore)
+	e.Logger = zap.New(newCore)
+	e.TaskManager.Logger = e.Logger
+
+	return logFile, nil
+}
+
+func closeQueryLogWriter(fd uintptr, e *Executor) {
+	// Uses the file descriptor pointer since we can remove or rename files
+	if err := syscall.Fsync(int(fd)); err != nil {
+		e.Logger.Error("failed to sync log file", zap.Error(err))
+	}
+	if err := syscall.Close(int(fd)); err != nil {
+		e.Logger.Error("failed to close log file", zap.Error(err))
+		return
+	}
+}
+
+func (e *Executor) WithLogWriter(log *zap.Logger, path string) {
+	var file *os.File
+	var err error
+
+	file, err = initQueryLogWriter(log, e, path)
+	if err != nil {
+		e.Logger.Error("failed to open log file", zap.Error(err))
+		return
+	}
+
+	go func() {
+		watcher, err := fsnotify.NewWatcher()
+		if err != nil {
+			e.Logger.Error("failed to create log file watcher", zap.Error(err))
+			closeQueryLogWriter(file.Fd(), e)
+			return
+		}
+
+		err = watcher.Add(path)
+		if err != nil {
+			e.Logger.Error("failed to watch log file", zap.Error(err))
+			closeQueryLogWriter(file.Fd(), e)
+			return
+		}
+		defer watcher.Close()
+
+		for {
+			select {
```

Comment:
I think we need an exit case here; a `context` that is cancelled on `influxd` shutdown, or some similar mechanism that closes and cleans up the watcher and the log.

Reply:
Otherwise we may not get the Sync() and Close() on the file, I believe.